### PR TITLE
Assignment 3 Pieter

### PR DESCRIPTION
**Pull request by lsobtafo**

### CODE REVIEW

**Comment by Tom-Rohr** [comment]
I'll review this.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
+﻿using System;
+namespace CanHazFunny
+{
+	public class Jester
+	{
+		private readonly IJokeService _jokeService;
```
Comment (Tom-Rohr) [inline CanHazFunny/CanHazFunny/Jester.cs:6]: I think the naming convention for class fields is pascal case. So _JokeService and _JokeOutput might be a good change

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        {
+            if (jokeService == null)
+            {
+                throw new System.ArgumentNullException(nameof(jokeService));
+            }
+            if (jokeOutput == null)
+            {
+                throw new System.ArgumentNullException(nameof(jokeOutput));
+            }
+            _jokeService = jokeService;
```
Comment (Tom-Rohr) [inline CanHazFunny/CanHazFunny/Jester.cs:18]: Same thing as lines 6 and 7. Pascal case for class field names.

File: CanHazFunny/CanHazFunny/IJokeService.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿using System;
+namespace CanHazFunny
+{
+	public interface IJokeService
```
Comment (Tom-Rohr) [inline CanHazFunny/CanHazFunny/IJokeService.cs:4]: lines 4 and 5 indented too far.

**Review by Tom-Rohr — COMMENTED** [review]
There is no class that implements your IJokeServiceOutput interface, so there is no way to instantiate a Jester to call TellJoke() for testing. Some other problems I found in the code had to do with formatting and naming convention. Needs unit tests still.

-Create *two* interfaces. One to represent the `JokeService` and one interface to represent outputting the joke to the screen (`Console.WriteLine`). ✔
- The `JokeService` will need to have the interface applied to it. ✔
- Create an implementation of the output interface that writes out the joke to the console. ✔
- Implement the `Jester` class. It should take in both interfaces as dependencies. These dependencies should be null checked. ❌
- The `Jester` class `TellJoke()` method should retrieve a joke from the `JokeService`. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ❌
- Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ❌

## Fundamentals

- Be sure you enable:
  - Nullability for all projects  ✔
  - Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ✔
  - Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ❌
- **Ensure there are no errors or warnings (including code analysis warnings)** ✔
- **All of the above should be unit tested** ❌
- **Choose simplicity over complexity** ✔
  
## Extra Credit

- Unit test your implementation that writes the joke out to the screen. How hard could it be to unit test a single line method ;)?❌
- The Geek jokes API that is being used can also return jokes in a JSON format. Update the `JokeService` to retrieve jokes using JSON. `GetJoke` should still return a string. ❌

File: CanHazFunny/CanHazFunny/IJokeService.cs (hunk 3 above)
Comment (eroadhouse) [reply CanHazFunny/CanHazFunny/IJokeService.cs:4]: I second this

File: CanHazFunny/CanHazFunny/IJokeServiceDisplay.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+namespace CanHazFunny
+{
+	public interface IJokeServiceDisplay
```
[the remark this answers is not in the document]
Comment (eroadhouse) [reply CanHazFunny/CanHazFunny/IJokeServiceDisplay.cs:4]: I second this

File: CanHazFunny/CanHazFunny/IJokeServiceOuput.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+namespace CanHazFunny
+{
+	public interface IJokeServiceOutput :IJokeServiceDisplay
```
Comment (eroadhouse) [inline CanHazFunny/CanHazFunny/IJokeServiceOuput.cs:4]: This class should also not be an interface.  The assignment says to create _two_ interfaces, not three. This class should just have one method that implements the Display method from IJokeServiceDisplay

File: CanHazFunny/CanHazFunny/IJokeServiceDisplay.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+	{
+        public void Display(string message)
+        {
+            Console.WriteLine(message);
```
Comment (eroadhouse) [inline CanHazFunny/CanHazFunny/IJokeServiceDisplay.cs:8]: Remember we want to leave implementation to classes the implement interfaces, not the interfaces themselves.

**Review by eroadhouse — COMMENTED** [review]
I think you just need to make IJokeServiceOutput not an interface and rather just a method that implements IJokeServiceDisplay. Then you will actually be able to instantiate Jester to be able to test TellJoke() if you change the constructor in Jester. Then obviously you'll need to unit test them

Create two interfaces. One to represent the JokeService and one interface to represent outputting the joke to the screen (Console.WriteLine). ✔ (Implemented 3 instead of 2)
The JokeService will need to have the interface applied to it. ✔
Create an implementation of the output interface that writes out the joke to the console. ✔
Implement the Jester class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
The Jester class TellJoke() method should retrieve a joke from the JokeService. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ❌
Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ❌
Fundamentals
Be sure you enable:
Nullability for all projects ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
For this assignment, always use Assert.AreEqual<T>() (the generic version) ❌
Ensure there are no errors or warnings (including code analysis warnings) ✔
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

**Comment by JLee35** [comment]
Unfortunately, it looks like a decent amount of the code was missing or incorrect. There are several reviews of your code made by your peers addressing these issues, but I don't see any responses or updates to fix them. First, while C# interfaces can have implementations, the goal of this assignment was to practice our polymorphism skills, and have concrete classes define the behavior of the methods in the interfaces they were implementing. This was not met, and all the functionality was in the interfaces themselves. Another primary goal in each assignment is to learn the importance of testing (hence why Mark teaches Test First Development). There were no tests, however, which is a major part of the grade. If you have any questions, please reach out to the instructor email.

**Pull request closed without merging**